test/MediatR.Tests/GenericRequestHandlerTests.cs

```diff
@@ -0,0 +1,190 @@
+﻿using MediatR.Extensions.Microsoft.DependencyInjection.Tests;
+using Microsoft.Extensions.DependencyInjection;
+using Shouldly;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection.Emit;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using System.Reflection.PortableExecutable;
+using MediatR.Tests.MicrosoftExtensionsDI;
+
+namespace MediatR.Tests
+{
+    public class GenericRequestHandlerTests : BaseGenericRequestHandlerTests
+    {
+
+        [Theory]
+        [InlineData(9, 3, 3)]
+        [InlineData(10, 4, 4)]
+        [InlineData(1, 1, 1)]
+        [InlineData(50, 3, 3)]
+        public void ShouldResolveAllCombinationsOfGenericHandler(int numberOfClasses, int numberOfInterfaces, int numberOfTypeParameters)

```

Comment:
What is this test doing?  This is not very readable

Reply:
the test is taking as parameters a variable number of classes, interfaces, and type parameters.

The classes and interfaces are generated like so:

let's say that the number of classes is 3 and the number of interfaces is 2. The following classes and interfaces will be generated.

```csharp
//number of interfaces: 2
public interface ITestInterface1 {}
public interface ITestInterface2 {}

//number of classes: 3
//interface implementation is spread out evenly among the classes with rollover
public class TestClass1 : ITestInterface1 {}
public class TestClass2 : ITestInterface2 {}
public class TestClass3 : ITestInterface1 {} 
```
If the number of generic parameters is 2 then the following generic request / handler is made

```csharp
public DynamicRequest<T1, T2> : IRequest
    where T1 : ITestInterface1
    where T2 : ITestInterface2
{ }

public DynamicRequestHandler<T1, T2> : IRequestHandler<DynamicRequest<T1, T2>>
   where T1 : ITestInterface1
   where T2 : ITestInterface2
{
    public Task Handle(DynamicRequest<T1, T2> request, CancellationToken cancelationToken)
    {
         return Task.CompletedTask;
    }
}
```
The last thing the test does is generate all combinations for concrete DynamicRequest implementations. It then loops through each combination and creates a test that attempts to retrieve the service and ensure that it is registerd (not null);

```csharp
provider.GetService<IRequestHandler<TestClass1, TestClass2>>().ShouldNotBeNull();
provider.GetService<IRequestHandler<TestClass3, TestClass2>>().ShouldNotBeNull();
```

Reply:
resolving since there is no suggestion being mentioned here only a clarification.